authenticator.go

```diff
@@ -0,0 +1,87 @@
+package main
+
+import (
+	"bytes"
+	"encoding/hex"
+	"io/ioutil"
+	"net/http"
+
+	p2pcrypto "github.com/libp2p/go-libp2p-crypto"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type Authenticator interface {
+	AuthenticateRequest(r *http.Request) bool
+}
+
+type P2PAuthenticator struct {
+	// B64 encoded public keys of identities whose logs are allowed to be
+	// ingested
+	allowedPubKeys []string
+
+	pubKeyMap map[string]p2pcrypto.PubKey
+}
+
+func NewP2PAuthenticator(allowedPubKeys []string) (*P2PAuthenticator, error) {
+	pubKeys := make(map[string]p2pcrypto.PubKey)
+
+	for _, encodedPubKey := range allowedPubKeys {
+		pubBytes, err := hex.DecodeString(encodedPubKey)
+		if err != nil {
+			return nil, err
+		}
+
+		pubKey, err := p2pcrypto.UnmarshalSecp256k1PublicKey(pubBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		pubKeys[encodedPubKey] = pubKey
+	}
+
+	return &P2PAuthenticator{
+		allowedPubKeys: allowedPubKeys,
+		pubKeyMap:      pubKeys,
+	}, nil
+}
+
+// AuthenticateRequest checks the id of the forwarder and validates the
+// signature
+// TODO: Return error for more specific validation errors?
+func (p *P2PAuthenticator) AuthenticateRequest(r *http.Request) bool {
```

Comment:
It affords the caller more flexibility to return an error rather then a boolean since the caller can then choose what to do with it. They could log in a single place instead of multiple, and can still get a boolean if desired `didAuthenticate := err == nil`.

```
if err != nil {
  log.WithField("error", err).Error("AuthenticateRequest returned an error")
}
```